#Text data

In [1]:
text="""The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering. Because it has achieved significance within the past fifty years, Criteria Consideration G applies. Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA. Unlike the Mercury, Gemini, and Apollo programs, the SSP’s emphasis was on cost effectiveness and reusability, and eventually the construction of a space station. Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. She had the honor of being chosen as the Return to Flight vehicle after both the Challenger and Columbia accidents. Discovery was the first shuttle to fly with the redesigned SRBs, a result of the Challenger accident, and the first shuttle to fly with the Phase II and Block I SSME. Discovery also carried the Hubble Space Telescope to orbit and performed two of the five servicing missions to the observatory. She flew the first and last dedicated Department of Defense (DoD) missions, as well as the first unclassified defense-related mission. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle. She was the first orbiter to dock to the ISS, and the first to perform an exchange of a resident crew. Under Criterion C, Discovery is significant as a feat of engineering. According to Wayne Hale, a flight director from Johnson Space Center, the Space Shuttle orbiter represents a “huge technological leap from expendable rockets and capsules to a reusable, winged, hypersonic, cargo-carrying spacecraft.” Although her base structure followed a conventional aircraft design, she used advanced materials that both minimized her weight for cargo-carrying purposes and featured low thermal expansion ratios, which provided a stable base for her Thermal Protection System (TPS) materials. The Space Shuttle orbiter also featured the first reusable TPS; all previous spaceflight vehicles had a single-use, ablative heat shield. Other notable engineering achievements of the orbiter included the first reusable orbital propulsion system, and the first two-fault-tolerant Integrated Avionics System. As Hale stated, the Space Shuttle remains “the largest, fastest, winged hypersonic aircraft in history,” having regularly flown at twenty-five times the speed of sound."""

#SpaCy


In [2]:
import spacy
from spacy.lang.en.stop_words import STOP_WORDS as spacy_SW
from string import punctuation as punct
from heapq import nlargest

In [3]:
nlp=spacy.load("en_core_web_sm")
doc=nlp(text)

In [4]:
tokens=[token.text for token in doc]

In [5]:
sp_stopwords=list(spacy_SW)
len(sp_stopwords)

326

In [6]:
punctuation=punct
punctuation = punctuation + '\n'

In [7]:
word_frequencies = {}
for word in doc:
  if word.text.lower() not in sp_stopwords:
    if word.text.lower() not in punctuation:
      if word.text not in word_frequencies.keys():
        word_frequencies[word.text] = 1
      else:
        word_frequencies[word.text] += 1

In [8]:
max_frequency=max(word_frequencies.values())
for word in word_frequencies.keys():
  word_frequencies[word]=word_frequencies[word]/max_frequency

In [9]:
sentence_tokens=list(doc.sents)
len(sentence_tokens)

17

In [10]:
sentence_score={}
for sent in sentence_tokens:
  for word in sent:
    if word.text.lower() in word_frequencies.keys():
      if sent not in sentence_score.keys():
        sentence_score[sent] = word_frequencies[word.text.lower()]
      else:
        sentence_score[sent] += word_frequencies[word.text.lower()]

In [11]:
sentence_score

{The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering.: 2.5000000000000004,
 Because it has achieved significance within the past fifty years, Criteria Consideration G applies.: 0.5,
 Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA.: 2.9000000000000004,
 Unlike the Mercury, Gemini, and Apollo programs, the SSP’s emphasis was on cost effectiveness and reusability, and eventually the construction of a space station.: 1.3,
 Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four or

In [12]:
from heapq import nlargest

In [13]:
select_length = int(len(sentence_tokens))
select_length

17

In [14]:
summary = nlargest(select_length, sentence_score, key = sentence_score.get)

In [15]:
summary

[According to Wayne Hale, a flight director from Johnson Space Center, the Space Shuttle orbiter represents a “huge technological leap from expendable rockets and capsules to a reusable, winged, hypersonic, cargo-carrying spacecraft.”,
 Although her base structure followed a conventional aircraft design, she used advanced materials that both minimized her weight for cargo-carrying purposes and featured low thermal expansion ratios, which provided a stable base for her Thermal Protection System (TPS) materials.,
 In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle.,
 Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions.,
 Under Criterion A, Discovery is significant as the oldest of the three extant orbiter ve

# NLTK

In [16]:
import nltk
nltk.download('punkt')
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize, sent_tokenize

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [17]:
nltk_tokens = word_tokenize(text)

In [18]:
sentences_list=sent_tokenize(text)
len(sentences_list)

16

In [19]:
nltk_stop_words=set(stopwords.words('english'))
len(nltk_stop_words)

179

In [20]:
nltk_word_frequencies = {}
for word in nltk_tokens:
  if word.lower() not in nltk_stop_words:
    if word.lower() not in punctuation:
      if word not in nltk_word_frequencies.keys():
        nltk_word_frequencies[word] = 1
      else:
        nltk_word_frequencies[word] += 1

In [21]:
# normalized frequency
nltk_max_frequency=max(nltk_word_frequencies.values())
for word in nltk_word_frequencies.keys():
  nltk_word_frequencies[word]=nltk_word_frequencies[word]/nltk_max_frequency

In [22]:
nltk_sentence_score={}
for sent in sentences_list:
  for word in word_tokenize(sent.lower()):
    if word.lower() in nltk_word_frequencies.keys():
      if sent not in nltk_sentence_score.keys():
        nltk_sentence_score[sent] = nltk_word_frequencies[word.lower()]
      else:
        nltk_sentence_score[sent] += nltk_word_frequencies[word.lower()]

In [23]:
len(nltk_sentence_score)

16

In [24]:
nltk_sentence_score

{'The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering.': 2.4000000000000004,
 'Because it has achieved significance within the past fifty years, Criteria Consideration G applies.': 0.7,
 'Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA.': 3.3000000000000007,
 'Unlike the Mercury, Gemini, and Apollo programs, the SSP’s emphasis was on cost effectiveness and reusability, and eventually the construction of a space station.': 1.4,
 'Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the othe

In [25]:
nltk_select_length = int(len(sentences_list))
nltk_select_length

16

In [26]:
nltk_summary = nlargest(nltk_select_length, nltk_sentence_score, key = nltk_sentence_score.get)

In [27]:
nltk_summary

['According to Wayne Hale, a flight director from Johnson Space Center, the Space Shuttle orbiter represents a “huge technological leap from expendable rockets and capsules to a reusable, winged, hypersonic, cargo-carrying spacecraft.” Although her base structure followed a conventional aircraft design, she used advanced materials that both minimized her weight for cargo-carrying purposes and featured low thermal expansion ratios, which provided a stable base for her Thermal Protection System (TPS) materials.',
 'Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions.',
 'Other notable engineering achievements of the orbiter included the first reusable orbital propulsion system, and the first two-fault-tolerant Integrated Avionics System.',
 'The Space Shuttle orbiter also featured the first reusable TPS; all previous spaceflight vehicles had a sin

#Conclusion

В цілому бібліотеки відпрацювали добре. Перші два реченні в обох моделях співпадають, хоча nltk не розбила достатнім чином перше речення тому включила два речення в одне. Третє речення в моделя не співпало,але четверте речення в моделі Spacy - це є третє в NLTK.

Бібліотека SpaCy розбила текст на 17 речень.

Бібліотека NLTK розбила текст на 16 речень.

Очищення тексту від пунктуації для обох моделей відбувалося з застосуванням складу знаків з пакету string +/n.

При обробці тексту SpaCy, було використано список з 326 stop_words, в той час NLTK мала в переліку 179 stop_words

На мою суб*єктивну думку з цим тестом Spacy справилася краще.